### DISCUSSION TITLE
Jest next/image: Cannot destructure property `deviceSizes` of 'undefined' or 'null'

### DISCUSSION THREAD

**Discussion by willemliu** [vercel/next.js#18373] [Help] [answered]
I'm trying out the new next/image component and replaced an img element. However when running my unit tests I encounter this error:

![image](https://user-images.githubusercontent.com/974906/97416809-359ed800-1907-11eb-9ee1-d9c56852afbf.png)

Anyone has any idea how to fix this?

**Comment by lightsound** [comment]
I have the same problem.

**Comment by willemliufdmg** [comment, marked as answer]
I've managed to create a work-around by adding the following `process.env` in a Jest [setupFile](https://jestjs.io/docs/en/configuration#setupfiles-array):
```
process.env = {
    ...process.env,
    __NEXT_IMAGE_OPTS: {
        deviceSizes: [320, 420, 768, 1024, 1200],
        imageSizes: [],
        domains: ['images.example.com'],
        path: '/_next/image',
        loader: 'default',
    },
};
```
Just remember to add your specific domains to the domains list.

**Reply by willemliufdmg** [reply]
I came with this work-around after some digging in the Next.js source and saw this:
```
const imageData: ImageData = process.env.__NEXT_IMAGE_OPTS as any
const {
  deviceSizes: configDeviceSizes,
  imageSizes: configImageSizes,
  loader: configLoader,
  path: configPath,
  domains: configDomains,
} = imageData
// sort smallest to largest
configDeviceSizes.sort((a, b) => a - b)
configImageSizes.sort((a, b) => a - b)
```

**Reply by SergiuszMinsk** [reply]
Seems it doesn't work for me. I added ```proccess evn``` to my ```jest.config.js``` and it can't find ```configDeviceSizes```  property. All tests failed with the new ```Image``` component.

next.js - v.10
jest - v.24.9.0
ts-jest - v.26.4.3

jest.config.js:

```
process.env = {
    ...process.env,
    __NEXT_IMAGE_OPTS: {
        deviceSizes: [320, 420, 768, 1024, 1200],
        imageSizes: [],
        domains: ["static.ghost.org"],
        path: "/_next/image",
        loader: "default",
    },
};
module.exports = {
    moduleFileExtensions: ["js", "jsx", "ts", "tsx"],
    globals: {
        "ts-jest": {
            tsConfig: "tsconfig.jest.json",
            babelConfig: true,
            diagnostics: false,
        },
    },
    coveragePathIgnorePatterns: ["/node_modules/", "/server/"],
    setupFilesAfterEnv: ["<rootDir>/test/jest.setup.ts"],
    coverageReporters: ["json", "lcov", "text", "text-summary"],
    moduleNameMapper: {
        "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$":
            "<rootDir>/test/mocks.ts",
        "\\.(css|less|scss|html)$": "<rootDir>/test/mocks.ts",
        "^@(Test)(.*)$": "<rootDir>/test/$2",
        "^@(Server)(.*)$": "<rootDir>/server/$2",
        "^@([A-Z].*)$": "<rootDir>/src/$1",
    },
    preset: "ts-jest",
    testEnvironment: "node",
};
```

Result:

<img width="720" alt="Screen Shot 2020-10-31 at 11 28 43 AM" src="https://user-images.githubusercontent.com/18442677/97776897-587df600-1b6c-11eb-8885-fb93d085c046.png">

**Reply by willemliu** [reply]
You need to create a setupTest.js file to which you need to point to from your jest.config.js file with the setupFiles property like seen here: https://github.com/FDMediagroep/fd-design-system/blob/develop/jest.config.js

In the setupTest.js file you then put the env vars like here: https://github.com/FDMediagroep/fd-design-system/blob/develop/jest/setupTest.js

**Reply by SergiuszMinsk** [reply]
thanks! now it works

**Reply by brunonardini** [reply]
@willemliu works for me too, thanks!

**Reply by avrcoelho** [reply]
Thaks

**Comment by Ihatetomatoes** [comment]
Same here, the workaround worked though. Thanks @willemliufdmg

**Comment by weyert** [comment]
Anyone been able to get this Image component working with Storybook? I think you get a similar error when using <Image />-component

**Reply by tifa2UP** [reply]
Haven't been able to get it running with storybook as well

**Reply by aippili-asp** [reply]
Same for me unable to run it in story book! @tifa2UP @weyert Did u find any workaround for this issue?

**Comment by tenhilltree** [comment]
I came across the same problem. I found the reason is that  I set the external config in my next.config.js.

![image](https://user-images.githubusercontent.com/13428695/98896316-a5a28600-24e3-11eb-9cfe-fb773feb26b7.png)

Then I  add the following code(from next.js's webpack-config.js),the problem is gone.

![image](https://user-images.githubusercontent.com/13428695/98896510-0af67700-24e4-11eb-9816-7057e566dbf1.png)
